common/scala/src/main/scala/whisk/common/tracing/OpenTracingProvider.scala

```diff
@@ -0,0 +1,195 @@
+
+/**
+ * OpenTracing based implementation for tracing
+ */
+class OpenTracer(val tracer: Tracer) extends WhiskTracer {
+
+  /**
+   * Start a Trace for given service.
+   *
+   * @param transactionId transactionId to which this Trace belongs.
+   * @return TracedRequest which provides details about current service being traced.
+   */
+  override def startSpan(logMarker: LogMarkerToken, transactionId: TransactionId): Unit = {
+
+    //initialize list for this transactionId
+    val spanList: List[WeakReference[Span]] = TracingCacheProvider.spanMap.get(transactionId.meta.id).getOrElse(Nil)
+
+    val spanBuilder = tracer
+      .buildSpan(logMarker.action)
+      .withTag("transactionId", transactionId.meta.id)
+
+    val activeSpan: Span = spanList match {
+      case Nil =>
+        TracingCacheProvider.contextMap
+          .get(transactionId.meta.id)
+          .map(spanBuilder.asChildOf(_).startActive(true).span())
+          .getOrElse(spanBuilder.ignoreActiveSpan().startActive(true).span())
+
+      case _ =>
+        spanList.head.get
+          .map(spanBuilder.asChildOf(_).startActive(true).span())
+          .getOrElse(spanBuilder.ignoreActiveSpan().startActive(true).span())
+
+    }
+    //add active span to list
+    TracingCacheProvider.spanMap.put(transactionId.meta.id, spanList.::(new WeakReference(activeSpan)))
+  }
+
+  /**
+   * Finish a Trace associated with given transactionId.
+   *
+   * @param transactionId
+   */
+  override def finishSpan(transactionId: TransactionId): Unit = {
+    clear(transactionId)
+  }
+
+  /**
+   * Register error
+   *
+   * @param transactionId
+   */
+  override def error(transactionId: TransactionId): Unit = {
+    clear(transactionId)
+  }
+
+  /**
+   * Get the current TraceContext which can be used for downstream services
+   *
+   * @param transactionId
+   * @return
+   */
+  override def getTraceContext(transactionId: TransactionId): Option[Map[String, String]] = {
+    TracingCacheProvider.spanMap.get(transactionId.meta.id) match {
+      case Some(spanList) => {
+        val map: mutable.Map[String, String] = mutable.HashMap()
+        //inject latest span context in map
+        spanList.head.get match {
+          case Some(span) => {
+            tracer.inject(span.context(), Format.Builtin.TEXT_MAP, new TextMapInjectAdapter(map.asJava))
+            Some(map.toMap)
+          }
+          case _ => None
+        }
+      }
+      case None => None
+    }
+  }
+
+  /**
+   * Get the current TraceContext which can be used for downstream services
+   *
+   * @param transactionId
+   * @return
+   */
+  override def setTraceContext(transactionId: TransactionId, context: Option[Map[String, String]]) = {
+    context match {
+      case Some(scalaMap) => {
+        val ctx: SpanContext = tracer.extract(Format.Builtin.TEXT_MAP, new TextMapExtractAdapter(scalaMap.asJava))
+        TracingCacheProvider.contextMap.put(transactionId.meta.id, ctx)
+      }
+      case None =>
+    }
```

Comment:
Please use Option combinators here as well:

```scala
context.foreach { scalaMap =>
  val ctx: SpanContext = tracer.extract(Format.Builtin.TEXT_MAP, new TextMapExtractAdapter(scalaMap.asJava))
  TracingCacheProvider.contextMap.put(transactionId.meta.id, ctx)
}

```

This is a great resource of how to idiomatically work with Options: http://blog.tmorris.net/posts/scalaoption-cheat-sheet/